sound/soc/sof/intel/hda-dai.c

```diff
@@ -568,6 +568,11 @@ int sdw_hda_dai_hw_params(struct snd_pcm_substream *substream,
 	 */
 	for_each_rtd_cpu_dais(rtd, i, dai) {
 		w = snd_soc_dai_get_widget(dai, substream->stream);
+		if (!w) {
+			dev_err(cpu_dai->dev,
+				"cpu dai widget not found, check amp link num in the topology\n");
+			return -EINVAL;
```

Comment:
there is no way to catch this during probe?

Reply:
I don't think we can catch this during probe since we need `substream->stream`. What do you think? @ranj063

Reply:
There's one possibility of making sure that all cpu DAI's associated with all runtimes in the card have their respective stream widgets set. What we can do is in the topology complete op, we can iterate though all of the rtd's and check if they're set and fail if they aren't. Worth a try. Something like this in sof_complete()

```
struct snd_soc_card *card = scomp->card;
list_for_each_entry(rtd, &card->rtd_list, list)
		for_each_rtd_cpu_dais(rtd, i, cpu_dai)
			if (!snd_soc_dai_get_widget(cpu_dai, stream))
                              return -EINVAL;
```

Reply:
Right, I can add
```c
        list_for_each_entry(rtd, &card->rtd_list, list)
                for_each_rtd_cpu_dais(rtd, i, cpu_dai)
                        if (!snd_soc_dai_get_widget(cpu_dai, SNDRV_PCM_STREAM_PLAYBACK) &&
                            !snd_soc_dai_get_widget(cpu_dai, SNDRV_PCM_STREAM_CAPTURE)) {
                                dev_err(sdev->dev, "no associated widgets for dai %s\n",
                                        cpu_dai->name);
                                return -EINVAL;
                        }
```
in sof_complete().
However, it will not return error if a dai has both directions and one direction isn't associated with a widget.

Reply:
so if we know what directions the CPU DAI has set, can we not check both or either based on that?

Reply:
The main issue is that the dai links may not always be used by the topology. I.e. topology could be a subset of the dai links. In that case the cpu dai will not associate with any widget and it is still valid.

Reply:
fair enough. this change looks good then